File: Objective-C/Demo/Test/TableViewComponent/FLTableComponentController.m
//
//  HYTableComponentController.m
//  live
//
//  Created by 孔凡列 on 2017/6/26.
//  Copyright © 2017年 YY Inc. All rights reserved.
//

#import "FLTableComponentController.h"

@interface FLTableComponentController ()

@property (nonatomic, strong) FLTableViewHandler *handler;

@property (nonatomic, strong) UITableView *tableView;

@end

@implementation FLTableComponentController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.view addSubview:self.tableView];
}

#pragma mark Public Method

- (CGRect)customTableViewBounds {
    return self.view.bounds;
}

- (UITableViewStyle)customTableViewStyle {
    return UITableViewStylePlain;
}

#pragma mark Setter & Getter

- (void)setComponents:(NSArray<FLTableComponent *> *)components {
    _components = components;
    self.handler.components = components;
}

- (UITableView *)tableView {
    if (_tableView == nil) {
        _tableView = [[UITableView alloc] initWithFrame:[self customTableViewBounds] style:[self customTableViewStyle]];
    }
    return _tableView;
}

- (FLTableViewHandler *)handler {
    if (_handler == nil) {
        _handler = [[FLTableViewHandler alloc] init];
    }
    return _handler;
}

@end

File: FLTableComponent/FLTableComponentController.swift
//
//  FLTableComponentController.swift
//  FLComponentDemo
//
//  Created by gitKong on 2017/5/11.
//  Copyright © 2017年 gitKong. All rights reserved.
//

import UIKit

class FLTableComponentController: UIViewController {
    
    private(set) var handler : FLTableViewHandler = FLTableViewHandler()
    
    lazy var tableView : UITableView = {
        let tableView : UITableView = UITableView.init(frame: self.customRect, style: self.tableViewStyle)
        return tableView
    }()
    
    var components : Array<FLTableBaseComponent> = [] {
        didSet {
            handler.components = components
        }
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        tableView.frame = self.customRect
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white
        handler.delegate = self
        view.addSubview(tableView)
        
        self.tableView.tableHeaderView = headerView(of: tableView)
        self.tableView.tableFooterView = footerView(of: tableView)
    }
    
    func headerView(of tableView : UITableView) -> UIView? {
        return nil
    }
    
    func footerView(of tableView : UITableView) -> UIView? {
        return nil
    }
}

extension FLTableComponentController : FLTableComponentConfiguration {
    
    var tableViewStyle: UITableViewStyle {
        return UITableViewStyle.plain
    }
    
    var customRect: CGRect {
        return self.view.bounds
    }
    
    func reloadComponent() {
        handler.reloadComponents()
    }
}

extension FLTableComponentController : FLTableViewHandlerDelegate {
    
    func tableViewDidClick(_ handler: FLTableViewHandler, cellAt indexPath: IndexPath) {
        // subclass override it
    }
    
    func tableViewDidClick(_ handler: FLTableViewHandler, headerAt section: NSInteger) {
        // subclass override it
    }
    
    func tableViewDidClick(_ handler: FLTableViewHandler, footerAt section: NSInteger) {
        // subclass override it
    }
}




